### DISCUSSION TITLE
Best practice to command u8g2.sendBuffer()

### DISCUSSION THREAD

**Discussion by ParallelTwin-Rider** [olikraus/u8g2#1825] [Q&A]
I appreciate for extreme efforts to provide u8g2.   Using the contribution, I am developing DIY rally computer.
2.4" 128 x 64 I2C SSD1309 displays various information such as distance and directions.  The number of items is 20.

I would like to ask the forum what is the best practice timing in the code to command u8g2.sendBuffer.

For example, there are items which obtain sensor information and figure out the value to show.  Current code is commanding everytime when the value is decided.  But it is also possible to command once in the end of loop.  In fact, video shooting the display looks different flicker.   My nature of this question is to shoot beautiful video without or less flicker by action cam.  So I am believing so far that I should optimize the code.
![post_pic](https://user-images.githubusercontent.com/100686546/160972171-c20096e3-84dd-4a65-9574-be40fc637e07.jpeg)

**Comment by olikraus** [comment]
Nice work, looks cool.

If flicker is an issue, then maybe the display itself is the root cause for the flicker. OLED displays do refresh with a fixed frame rate. For the SSD1309 sections 8.3 and 10.16 of the SSD1309 datasheet describe how this frame rate is derived and influenced: We can use the 0xd5 command to change the frame rate itself:
Quote from the datasheet:
> 10.16 Set Display Clock Divide Ratio/ Oscillator Frequency (D5h)
This command consists of two functions:
• Display Clock Divide Ratio (D) (A[3:0])
Set the divide ratio to generate DCLK (Display Clock) from CLK. The divide ratio is from 1 to 16,
with reset value = 0000b. Please refer to section 8.3 for the details relationship of DCLK and CLK.
• Oscillator Frequency (A[7:4])
Program the oscillator frequency Fosc that is the source of CLK if CLS pin is pulled high. The 4-bit
value results in 16 different frequency settings available as shown below. The default setting is
1000b.

Unfortunately the mentioned timing table for A[7:4] is not listed in the datasheet, but you could still try different values and check whether it works better with your cam. You can change the internal configuration of the display with the u8g2 sendF command (https://github.com/olikraus/u8g2/wiki/u8g2reference#sendf).
The default value seems to be:
```
u8g2.sendF("ca", 0xd5, 0x80);
```
The highest possible frame rate would be generated with 
```
u8g2.sendF("ca", 0xd5, 0x00);
```
This value may not work at all and even if it works, it might influence brightness, so maybe test
```
u8g2.sendF("ca", 0xd5, 0x70);
```
and decrese the highest 4 bit (=Oscillator Frequency) of the argument step by step:
```
u8g2.sendF("ca", 0xd5, 0x60);
```

You must use `u8g2.sendF()` after calling `u8g2.begin()`.

**Reply by ParallelTwin-Rider** [reply]
So wonderful to hear the reply above so promtly．I tried u8g2.sendF with these mentioned values, then these encoding were successsful and display can show strings.  Significant cam flicker will come up under sunshine due to automatice shutter speed adjustment.  So I will try them under perceived condition and share the result as a bit of contribution to the world.

**Reply by ParallelTwin-Rider** [reply]
Hello olikraus,
I tested with cam, and the result seems not good throughout with the  various values.   Under sunlight, cam changes shutter speed automatically to 1/1000, and display cannot be visible (only bright angled line, left up).   Then I switched to manual shutter speed, different resolution and fps, changing to various combination of your suggested value.   The suggested vaule is effective at inside of room.   However, under sunlight, the optimized shutter speed was 1/80 to take the display clearly.  But 1/80 even with auto ISO cannnot clearly take other than display.  So I guess this is a limitation of OLED in case of SSD1309 because of the fixed frame rate.   I have tried TFT/PSI display.  They did not bring these issues, but too dark under sunlight.  High bright TFT/PSI display is expensive though, so I selected OLED for my project as sunlight readable.   The function as rally computer and display is so readable.   Only my regret is that I cannot share the joy of this device and 'creation of things' for respective objectives and wishes of people.  Anyhow, I appreciate for your kind assistance and huge efforts on u8g2, which cultivate careative world.  Thank you so much.

**Comment by olikraus** [comment]
I assume the frame rate of the display will not be faster than 100Hz or so. So it will always be difficult to use 1/1000 shutter speed. Sounds like a unsolvable problem.

**Reply by ParallelTwin-Rider** [reply]
Yes that is the answer.
Now I’m exploring two directions.  One is to customize purchased PSI (200cd) to add more backlight LED.  Another is to use e-paper.  Monochrome e-paper can react 0.2 sec for partial update.  Still under research and inquiry to the manufacturer.

**Reply by ParallelTwin-Rider** [reply]
Sorry for bothering you by many messages.  
I found this device.   https://www.lcd-module.de/eng/pdf/grafik/dogm128e.pdf
This is monochrome LCD with ST7565R SPI.  I believe u8g2 is able to control this display.
If you could provide any comments or suggestion?

**Reply by olikraus** [reply]
Yes, indeed,  LCDs might be an option. They also use about 100 Hz frame rate, but the pixel values change much slower. ST7565 is supported by u8g2.

**Reply by olikraus** [reply]
BTW your LCD needs to be sunlight readable. Official terms are reflective LCD (best under sunlight) or transflective LCD (will be ok). Transmissive LCDs probably will not be readable under sunlight.

**Reply by ParallelTwin-Rider** [reply]
Hello olikraus,
Appreciate for your comment.  I will look for the module.
I uploaded you tube video as the test of adjustment. 
https://youtu.be/0Blj92LYFew

**Reply by olikraus** [reply]
Super video documentation. Thanks for sharing. Like+
